plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+	m_fftBlockSize(FFT_BLOCK_SIZES[0]),
+	m_sampleRate(Engine::mixer()->processingSampleRate()),
+	m_framesFilledUp(0),
+	m_spectrumActive(false),
+	m_waterfallActive(false),
+	m_waterfallNotEmpty(0),
+	m_destroyed(false),
+	m_reallocating(false)
+{
+	m_fftWindow.resize(m_inBlockSize, 1.0);
+	precomputeWindow(m_fftWindow.data(), m_inBlockSize, BLACKMAN_HARRIS);
+
+	m_bufferL.resize(m_fftBlockSize, 0);
+	m_bufferR.resize(m_fftBlockSize, 0);
+	m_spectrumL = (fftwf_complex *) fftwf_malloc(binCount() * sizeof (fftwf_complex));
+	m_spectrumR = (fftwf_complex *) fftwf_malloc(binCount() * sizeof (fftwf_complex));
+	m_fftPlanL = fftwf_plan_dft_r2c_1d(m_fftBlockSize, m_bufferL.data(), m_spectrumL, FFTW_MEASURE);
+	m_fftPlanR = fftwf_plan_dft_r2c_1d(m_fftBlockSize, m_bufferR.data(), m_spectrumR, FFTW_MEASURE);
+
+	m_absSpectrumL.resize(binCount(), 0);
+	m_absSpectrumR.resize(binCount(), 0);
+	m_normSpectrumL.resize(binCount(), 0);
+	m_normSpectrumR.resize(binCount(), 0);
+
+	m_history.resize(binCount() * m_waterfallHeight * sizeof qRgb(0,0,0), 0);
+
+	clear();
+}
+
+
+SaProcessor::~SaProcessor()
+{
+	// take a note that destructor was called; see reallocateBuffers()
+	m_destroyed = true;
+
+	if (m_fftPlanL != NULL) {fftwf_destroy_plan(m_fftPlanL);}
+	if (m_fftPlanR != NULL) {fftwf_destroy_plan(m_fftPlanR);}
+	if (m_spectrumL != NULL) {fftwf_free(m_spectrumL);}
+	if (m_spectrumR != NULL) {fftwf_free(m_spectrumR);}
+
+	m_fftPlanL = NULL;
+	m_fftPlanR = NULL;
+	m_spectrumL = NULL;
+	m_spectrumR = NULL;
+}
+
+
+// Load a batch of data from LMMS; run FFT analysis if buffer is full enough.
+void SaProcessor::analyse(sampleFrame *in_buffer, const fpp_t frame_count)
+{
+	#ifdef SA_DEBUG
+		int start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+	#endif
+	// only take in data if any view is visible and not paused
+	if ((m_spectrumActive || m_waterfallActive) && !m_controls->m_pauseModel.value())
+	{
+		const bool stereo = m_controls->m_stereoModel.value();
+		fpp_t in_frame = 0;
+		while (in_frame < frame_count)
+		{
+			// fill sample buffers and check for zero input
+			bool block_empty = true;
+			for (; in_frame < frame_count && m_framesFilledUp < m_inBlockSize; in_frame++, m_framesFilledUp++)
+			{
+				if (stereo)
+				{
+					m_bufferL[m_framesFilledUp] = in_buffer[in_frame][0];
+					m_bufferR[m_framesFilledUp] = in_buffer[in_frame][1];
+				}
+				else
+				{
+					m_bufferL[m_framesFilledUp] = (in_buffer[in_frame][0] + in_buffer[in_frame][1]) * 0.5;
+					m_bufferR[m_framesFilledUp] = (in_buffer[in_frame][0] + in_buffer[in_frame][1]) * 0.5;
+				}
+				if (in_buffer[in_frame][0] != 0 || in_buffer[in_frame][1] != 0)
+				{
+					block_empty = false;
+				}
+			}
+	
+			// Run analysis only if buffers contain enough data.
+			// Also, to prevent audio interruption and a momentary GUI freeze,
+			// skip analysis if buffers are being reallocated.
+			if (m_framesFilledUp < m_inBlockSize || m_reallocating) {return;}
```

Comment:
> What happens if the audio thread has already passed this if statement, and now the reallocation thread starts to reallocate?

The processor will wait on m_dataAccess lock. This condition is here simply to prevent the analysis from waiting on on the lock when it could skip the block instead. So if the condition fails for some reason, the worst that should happen is that the reallocation time will be added to the FFT processing time.

Reply:
Ah right, I get it. In this case, you also don't need an atomic, since it's secured by `m_dataAccess`. You can resolve that comment.